Source/lighting.cpp

```diff
@@ -283,90 +290,47 @@ void DoUnVision(Point position, int nRadius)
 	}
 }
 
-void DoVision(Point position, int nRadius, MapExplorationType doautomap, bool visible)
+void DoVision(Point position, int radius, MapExplorationType doAutomap, bool visible)
 {
-	if (InDungeonBounds(position)) {
-		if (doautomap != MAP_EXP_NONE) {
-			if (dFlags[position.x][position.y] != DungeonFlag::None) {
-				SetAutomapView(position, doautomap);
-			}
-			dFlags[position.x][position.y] |= DungeonFlag::Explored;
-		}
-		if (visible) {
-			dFlags[position.x][position.y] |= DungeonFlag::Lit;
-		}
-		dFlags[position.x][position.y] |= DungeonFlag::Visible;
+	if (doAutomap != MAP_EXP_NONE) {
+		if (dFlags[position.x][position.y] != DungeonFlag::None)
+			SetAutomapView(position, doAutomap);
+		dFlags[position.x][position.y] |= DungeonFlag::Explored;
 	}
+	if (visible)
+		dFlags[position.x][position.y] |= DungeonFlag::Lit;
+	dFlags[position.x][position.y] |= DungeonFlag::Visible;
 
-	for (int v = 0; v < 4; v++) {
+	if (radius == 0)
+		return;
+
+	static const auto factors = CrawlFlips({ { 1, 1 } });
+	for (auto factor : factors) {
 		for (int j = 0; j < 23; j++) {
-			bool nBlockerFlag = false;
-			int nLineLen = 2 * (nRadius - RadiusAdj[j]);
-			for (int k = 0; k < nLineLen && !nBlockerFlag; k += 2) {
-				int x1adj = 0;
-				int x2adj = 0;
-				int y1adj = 0;
-				int y2adj = 0;
-				int nCrawlX = 0;
-				int nCrawlY = 0;
-				switch (v) {
-				case 0:
-					nCrawlX = position.x + VisionCrawlTable[j][k];
-					nCrawlY = position.y + VisionCrawlTable[j][k + 1];
-					if (VisionCrawlTable[j][k] > 0 && VisionCrawlTable[j][k + 1] > 0) {
-						x1adj = -1;
-						y2adj = -1;
-					}
-					break;
-				case 1:
-					nCrawlX = position.x - VisionCrawlTable[j][k];
-					nCrawlY = position.y - VisionCrawlTable[j][k + 1];
-					if (VisionCrawlTable[j][k] > 0 && VisionCrawlTable[j][k + 1] > 0) {
-						y1adj = 1;
-						x2adj = 1;
-					}
-					break;
-				case 2:
-					nCrawlX = position.x + VisionCrawlTable[j][k];
-					nCrawlY = position.y - VisionCrawlTable[j][k + 1];
-					if (VisionCrawlTable[j][k] > 0 && VisionCrawlTable[j][k + 1] > 0) {
-						x1adj = -1;
-						y2adj = 1;
-					}
+			int lineLen = radius - RadiusAdj[j];

```

Comment:
You can get rid of RadiusAdj if you make VisionCrawlTable an array of vectors. The last entries of rays 8 and 14 (`13, 8` and `8, 13` respectively) don't get used, so can drop those and all the `0, 0` entries to match vanilla behaviour. The two innermost loops can then be range-based loops like
```cpp
for (auto ray : VisionCrawlTable)
    for (Displacement offset : ray)
```

Reply:
Oh, looking at this again this is somewhat broken. the vision table is built using rays with magnitude 15, but when used with smaller radii the adjustment should be scaled to match. This becomes more noticeable when radius <= 8...

![image](https://user-images.githubusercontent.com/357684/176984116-d8c6bd98-f6cb-4e9f-95fb-809ba4b1ad1b.png)
This table shows the magnitude of each vector between the origin and the tile rounded to the nearest integer. The cells with a red background have a magnitude > 15, if the intent was always to round down then they shouldn't be included. If the intent was to fudge the boundary for a more pleasing shape, 11,11 should probably be excluded.

For an example of the scaling impacting what area gets lit: using a radius of 4 ends up completely skipping the diagonal ray, though there's enough duplication in the other rows of the table that doesn't matter. There are still gaps because of the adjustments applied to other rows, the two highlighted cells are never checked when they should be included:
![image](https://user-images.githubusercontent.com/357684/176984547-c8393845-caf4-4545-8fb4-1f32a2fcaa8c.png)

The "correct" adjustment would be something along the lines of
```cpp
size_t effectiveRadius = radius * ray.size() / 15; // using truncation towards zero
auto end = ray.begin() + effectiveRadius;
```

Reply:
oooor, how about making the lookup table have 15 entries for each ray. I'll put a comment on the table with a suggestion of how that'd look